src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,279 @@
+                           << __func__ << ": "
+
+namespace efs = std::experimental::filesystem;
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+ObjectCacheStore::ObjectCacheStore(CephContext *cct)
+      : m_cct(cct), m_rados(new librados::Rados()),
+        m_ioctx_map_lock("ceph::cache::ObjectCacheStore::m_ioctx_map_lock") {
+
+  object_cache_max_size =
+    m_cct->_conf.get_val<Option::size_t>("immutable_object_cache_max_size");
+
+  std::string cache_path =
+    m_cct->_conf.get_val<std::string>("immutable_object_cache_path");
+  m_cache_root_dir = cache_path + "/ceph_immutable_obj_cache/";
+
+  //TODO(): allow to set cache level
+  m_policy = new SimplePolicy(m_cct, object_cache_max_size/(4096*1024), 0.1);
+}
+
+ObjectCacheStore::~ObjectCacheStore() {
+  delete m_policy;
+  for (auto it: m_ioctx_map) {
+    delete it.second;
+  }
+}
+
+int ObjectCacheStore::init(bool reset) {
+  ldout(m_cct, 20) << dendl;
+
+  int ret = m_rados->init_with_context(m_cct);
+  if (ret < 0) {
+    lderr(m_cct) << "fail to init Ceph context" << dendl;
+    return ret;
+  }
+
+  ret = m_rados->connect();
+  if (ret < 0 ) {
+    lderr(m_cct) << "fail to connect to cluster" << dendl;
+    return ret;
+  }
+
+  //TODO(): fsck and reuse existing cache objects
+  if (reset) {
+    std::error_code ec;
+    if (efs::exists(m_cache_root_dir)) {
+      if (!efs::remove_all(m_cache_root_dir, ec)) {
+        lderr(m_cct) << "fail to remove old cache store: " << ec << dendl;
+	return -1;
+      }
+    }
+
+    if (!efs::create_directories(m_cache_root_dir, ec)) {
+        lderr(m_cct) << "fail to create cache store dir: " << ec << dendl;
+	return -1;
+    }
+  }
+
+  return ret;
+}
+
+int ObjectCacheStore::shutdown() {
+  ldout(m_cct, 20) << dendl;
+
+  m_rados->shutdown();
+  return 0;
+}
+
+int ObjectCacheStore::init_cache() {
+  ldout(m_cct, 20) << dendl;
+  std::string cache_dir = m_cache_root_dir;
+
+  int dir = m_dir_num - 1;
+  while (dir >= 0) {
+    efs::create_directories(cache_dir + "/" + std::to_string(dir));
+    dir --;
+  }
+  return 0;
+}
+
+int ObjectCacheStore::do_promote(std::string pool_nspace,
+                                  uint64_t pool_id, uint64_t snap_id,
+                                  std::string object_name) {
+  ldout(m_cct, 20) << "to promote object: " << object_name
+                   << " from pool id: " << pool_id
+                   << " namespace: " << pool_nspace
+                   << " snapshot: " << snap_id << dendl;
+
+  int ret = 0;
+  std::string cache_file_name = std::move(get_cache_file_name(pool_nspace,
+                                          pool_id, snap_id, object_name));
+  librados::IoCtx* ioctx = nullptr;
+  {
+    Mutex::Locker _locker(m_ioctx_map_lock);
+    if (m_ioctx_map.find(pool_id) == m_ioctx_map.end()) {
+      ioctx = new librados::IoCtx();
+      ret = m_rados->ioctx_create2(pool_id, *ioctx);
+      if (ret < 0) {
+        lderr(m_cct) << "fail to create ioctx" << dendl;
+        return ret;
```

Comment:
Nit: memory leak of ```ioctx``` -- but you really don't need to allocate them dynamically either since they internally use reference counting.